teps/0156-whenexpressions-in-step.md

```diff
@@ -0,0 +1,116 @@
+
+## Proposal
+We propose extending the Step definition in Tekton Tasks to include an optional when field, similar to the existing implementation for Tasks and Pipelines. The when field will accept a list of When Expressions that define the conditions for executing the Step.
+
+### When Expressions Evaluation
+
+When expressions are evaluated at the entrypointer level. This allows us to use step results produced from previous steps to determine if a step should be skipped. To indicate a skipped step, the exit code will be 0 (showing a step is complete) and the TerminationReason will be Skipped.
+
+### Variable Substitution 
+
+With the introduction of When Expressions at the Step level, it's important to clarify which fields can be referenced within these expressions for variable substitution.
+ - Task-based fields:
+   - Task Results: You can access the results of any previously executed Task within the same TaskRun using the syntax $(tasks.<taskName>.results.<resultName>).
+ - Workspaces: You can reference the names of workspaces declared in the Task using the syntax $(workspaces.<workspaceName>.bound). This allows you to check if a workspace is bound and make decisions based on its availability.
+ - Params: Access input parameters defined in the Task using the familiar syntax $(params.<paramName>). This enables conditional logic based on the values passed into the Task.
+ - Step Results: The results of previously executed steps within the same Task are accessible using $(steps.<stepName>.results.<resultName>). This is crucial for creating dynamic workflows where subsequent steps depend on the outcomes of earlier ones.
+ - Step Status: Access the reason why a step terminated with $(steps.<stepName>.status.terminationReason). This field can take on values like "Error", "Completed", "Skipped", "TimeOut", or "Cancelled", providing insights into the step's execution and enabling more refined decision-making within When Expressions.
+
+### Supported Expression Languages
+
+Just like when expressions in Task Level, when expressions in Step Level will also be supporting both Operator-based expressions and CEL(Common Expression Language) expressions. 
+**Examples**
+ - Operator-based when expressions in steps.
+```yaml
+  apiVersion: tekton.dev/v1
+  kind: TaskRun
+  metadata:
+    generateName: operator-based-
+  spec:
+    workspaces:
+      - name: custom
+        persistentVolumeClaim:
+        claimName: my-pvc-2
+    taskSpec: 
+      steps:
+        - name: produce-step
+          image: alpine
+          results:
+             - name: result2
+               type: string
+          script: |
+             echo -n "foo" | tee $(step.results.result2.path)
+        - name: run-based-on-step-results
+          image: alpine
+          script: |
+             echo "wooooooo"
+          when:
+             - input: "$(steps.produce-step.results.result2)"
```

Comment:
Are we not creating a DAG at `Step` level currently?

Reply:
steps are executed in-order, so they're naturally a DAG